auacm/app/modules/problem_manager/views.py

```diff
@@ -1,89 +1,341 @@
+    solved = session.query(Submission).\
+            filter(Submission.username == current_user.username).\
+            filter(Submission.result == "good").\
+            all()
+    solved_set = set()
+    for solve in solved:
+        solved_set.add(solve.pid)
+
+    for problem in session.query(Problem).all():
+        problems.append({
+            'pid': problem.pid,
+            'name': problem.name,
+            'shortname': problem.shortname,
+            'appeared': problem.appeared,
+            'difficulty': problem.difficulty,
+            'compRelease': problem.comp_release,
+            'added': problem.added,
+            'solved': problem.pid in solved_set,
+            'url': url_for_problem(problem)
+        })
+    return serve_response(problems)
+
+
+@app.route('/api/problems/create', methods=['POST'])
+@login_required
+def create_problem():
+    """Add a new problem to the database and data folder"""
+    # Admin check
+    if not current_user.admin == 1:
+        return serve_error('You must be an admin to create problems',
+                           response_code=401)
+
+    try:
+        # Convert the JSON to python array of dictionaries
+        cases = request.form['cases']
+        cases = loads(str(cases))
+        for case in cases:
+            if not 'input' in case or not 'output' in case:
+                return serve_error(
+                    'Sample case(s) were not formed correctly',
+                    response_code=400)
+
+        # Create the problem
+        title = request.form['title'][:32]
+        shortname = title.lower().replace(' ', '')
+        problem = Problem(
+            name=title,
+            shortname=shortname
+        )
+        if 'difficulty' in request.form:
+            problem.difficulty = request.form['difficulty']
+        if 'appeared_in' in request.form:
+            problem.appeared = request.form['appeared_in']
+
+        # Create the problem data and add it to the database
+        problem_data = ProblemData(
+            description=request.form['description'],
+            input_desc=request.form['input_description'],
+            output_desc=request.form['output_description']
+        )
+        if 'time_limit' in request.form:
+            problem_data.time_limit = request.form['time_limit']
+
+        # Create list of sample cases
+        case_num = 1
+        sample_cases = list()
+        for case in cases:
+            sample = SampleCase(
+                case_num=case_num,
+                input=case['input'],
+                output=case['output']
+            )
+            case_num += 1
+            sample_cases.append(sample)
+
+        in_file = zipfile.ZipFile(request.files['in_file'])
+        out_file = zipfile.ZipFile(request.files['out_file'])
+        sol_file = request.files['sol_file']
+
+    # If any required values were missing, serve an error
+    except KeyError as err:
+        return serve_error('Request header not found: ' + err[0],
```

Comment:
Technically, it's not a request header (that would imply that it's an HTTP header). I'd just say something like "'Error, form incomplete; missing ' + err[0]." Something along those lines.